### PR TITLE
Reserve promise in top level module

### PR DESCRIPTION
**Pull request by rbuckton**
This address #6068.

We have chosen to more closely align with ES2016 (ES7) and restrict the return types of async functions in ES6 (and later) to only return instances of the global `Promise` object. This has the following ramifications:

## Return Types
If an async function or method has a return type, it must be a type reference to global generic `Promise` type. This aligns with the same return type requirements of C#.

## User-supplied Promise constructors
We will no longer instantiate a user-supplied `Promise` constructor during the initialization of an async function. Allocation of the Promise still happens inside of the `__awaiter` helper, so users are encouraged to define their own `__awaiter` helper if they need to use a custom Promise. We may still support user-supplied Promise constructors for ES5/3 when if/when we add support for generators and async functions via further transpilation, as these targets do not support a native Promise implementation per their specifications.

## `Promise` reserved at the top-level of a module
We now reserve the identifier `Promise` for any declarations at the top-level of a module containing async functions or methods. This aligns with how we already reserve `require` and `exports` at the top-level of a module. Any user today that currently imports `Promise` from an external module that wishes to use async functions or methods in their codebase will need to rename or alias the local `Promise` declaration to avoid this error.

For example:
```ts
import * as Promise from "bluebird";
// error TS2529: Duplicate identifier 'Promise'. Compiler reserves name 'Promise' in the top level scope of a module containing async functions.

export async function fn() {
  await Promise.delay(10);
}

```
It is still acceptable to reference `Promise` in an expression, so this is perfectly acceptable:

```ts
async function fn() {
  await Promise.resolve(1); // 
}
```

### CODE REVIEW

File: tests/baselines/reference/asyncFunctionDeclaration15_es6.errors.txt [hunk 1]
```diff
@@ -1,12 +1,8 @@
-tests/cases/conformance/async/es6/functionDeclarations/asyncFunctionDeclaration15_es6.ts(6,16): error TS1055: Type '{}' is not a valid async function return type.
```
Comment (yuit) [inline tests/baselines/reference/asyncFunctionDeclaration15_es6.errors.txt]: Should this error be removed? it seems that this error `The return type of an async function or method must be the global Promise<T>.` replaces it

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -12907,6 +12954,25 @@ namespace ts {
...
+
+            // Uninstantiated modules shouldnt do this check
+            if (node.kind === SyntaxKind.ModuleDeclaration && getModuleInstanceState(node) !== ModuleInstanceState.Instantiated) {
+                return;
+            }
+
+            // In case of variable declaration, node.parent is variable statement so look at the variable statement's parent
+            const parent = getDeclarationContainer(node);
+            if (parent.kind === SyntaxKind.SourceFile && isExternalOrCommonJsModule(<SourceFile>parent) && parent.flags & NodeFlags.HasAsyncFunctions) {
+                // If the declaration happens to be in external module, report error that require and exports are reserved keywords
```
Comment (yuit) [inline src/compiler/checker.ts]: This comment doesn't seem correct

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -12495,6 +12509,33 @@ namespace ts {
         }
 
         /**
+         * Checks that the return type provided is an instantiation of the global Promise<T> type
+         * and returns the awaited type of the return type.
```
Comment (yuit) [inline src/compiler/checker.ts]: since it is already jsdoc. could you add the parameter annotation as well?

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -12508,6 +12549,11 @@ namespace ts {
           * callable `then` signature.
           */
         function checkAsyncFunctionReturnType(node: FunctionLikeDeclaration): Type {
+            if (languageVersion >= ScriptTarget.ES6) {
```
Comment (yuit) [inline src/compiler/checker.ts]: why >= ES6? I thought need to resolve global Promise when down-leveling?

File: src/compiler/emitter.ts [hunk 5]
```diff
@@ -320,7 +320,7 @@ var __param = (this && this.__param) || function (paramIndex, decorator) {
 
         const awaiterHelper = `
 var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
-    return new P(function (resolve, reject) {
+    return new (P || (P = Promise))(function (resolve, reject) {
```
Comment (yuit) [inline src/compiler/emitter.ts]: why can't use always just do `new Promise` here?

File: src/compiler/checker.ts (hunk 4 above)
Comment (yuit) [inline src/compiler/checker.ts]: @rbuckton  explains off-line. ES6 has native Promise. I misunderstood that fact

File: src/compiler/emitter.ts (hunk 5 above)
Comment (rbuckton) [inline src/compiler/emitter.ts]: We are leaving support for custom Promise return types for ES5 and earlier which do not have a global Promise, so we want to keep the same argument list. The default behavior now will be to use the Promise captured in the same scope as `__awaiter`. This is to avoid issues when creating closures in namespaces:

```ts
// input.ts
export namespace x {
  var Promise = "Not a Promise constructor.";
  async function f(): Promise { // references global Promise type
  }
}

// output.js
var __awaiter = ...; // captures global Promise here
export var x = (function (x) {
  var Promise = "Not a Promise constructor."; // local Promise var shadows global Promise
  function f() {
    // Before this change, we would capture Promise here, which would be the wrong Promise:
    // return __awaiter(this, void 0, Promise, function *() { });
    
    // After this change, we use the Promise captured at the top-level of the module.
    return __awaiter(this, void 0, void 0, function *() { });
  }
})(x || (x = {}));
```

File: src/compiler/checker.ts (hunk 4 above)
Comment (rbuckton) [inline src/compiler/checker.ts]: ES5 and earlier do not have a global Promise. We fall back to the existing behavior for ES5 as we plan to support async functions via tree transformation for target ES5 and users will need to be able to supply a Promise implementation.

**Comment by yuit** [comment]
:+1: after some comments

File: src/compiler/diagnosticMessages.json [hunk 6]
```diff
@@ -195,6 +195,10 @@
         "category": "Error",
         "code": 1063
     },
+    "The return type of an async function or method must be the global Promise<T>.": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: `global 'Promise<T>' type.`

File: src/compiler/diagnosticMessages.json [hunk 7]
```diff
@@ -1687,6 +1691,10 @@
         "category": "Error",
         "code": 2528
     },
+    "Duplicate identifier '{0}'. Compiler reserves name '{1}' in top level scope of a module containing async functions.": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: The compiler reserves '{1}' in the top level scope of of a module when using async functions.

File: tests/baselines/reference/asyncFunctionDeclaration15_es6.errors.txt (hunk 1 above)
Comment (rbuckton) [inline tests/baselines/reference/asyncFunctionDeclaration15_es6.errors.txt]: These errors will remain as we are keeping the logic that uses them for use with downlevel async functions for ES5 and earlier.

File: src/compiler/checker.ts (hunk 2 above)
Comment (rbuckton) [inline src/compiler/checker.ts]: Vestige of copy/paste, I'll correct the comment.

File: src/compiler/diagnosticMessages.json (hunk 6 above)
Comment (rbuckton) [inline src/compiler/diagnosticMessages.json]: I can add "type". This was paraphrased from the following error message in C#:
```
The return type of an async method must be void, Task, or Task<T>.
```

File: src/compiler/diagnosticMessages.json (hunk 7 above)
Comment (rbuckton) [inline src/compiler/diagnosticMessages.json]: This is modeled after TS2441: `Duplicate identifier '{0}'. Compiler reserves name '{1}' in top level scope of a module.` See: https://github.com/Microsoft/TypeScript/blob/master/src/compiler/diagnosticMessages.json#L1350

File: src/compiler/checker.ts [hunk 8]
```diff
@@ -12495,6 +12509,36 @@ namespace ts {
         }
 
         /**
+         * Checks that the return type provided is an instantiation of the global Promise<T> type
```
Comment (yortus) [inline src/compiler/checker.ts]: What is the reason for the compiler to reject type annotations that are known compatible with `Promise`? This won't catch any type errors, but rejects valid annotations. Generator functions handle this consistently IMO. Why not use the same approach here for consistency?

```typescript
function* bar() {}                  // Returns IterableIterator<any>
function* bar2(): any {}            // OK: compatible annotation
function* bar3(): Iterator<any> {}  // OK: comtatible annotation
function* bar4(): { next } {}       // OK: comtatible annotation

async function foo() {}                     // Returns Promise<void>
async function foo1(): any {}               // ERROR with this PR, but why?
async function foo2(): PromiseLike<any> {}  // ERROR with this PR. but why?
```

File: src/compiler/diagnosticMessages.json (hunk 7 above)
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: I think we should change that as well then.

File: src/compiler/checker.ts (hunk 8 above)
Comment (yortus) [inline src/compiler/checker.ts]: The `checker.ts` code for generators uses `checkTypeAssignableTo` to check the return type annotation ([L11625](https://github.com/Microsoft/TypeScript/blob/master/src/compiler/checker.ts#L11625)). Why is `checkTypeAssignableTo` not used for async functions?

Comment (rbuckton) [inline src/compiler/checker.ts]: This is the result of an offline discussion with @mhegazy. Supporting an assignable interface is not currently compatible with our goals for future down-level support for async functions for ES5. As a result, async functions will for the time being be more restrictive than generators with respect to the return type annotation.

File: src/compiler/diagnosticMessages.json (hunk 7 above)
Comment (rbuckton) [inline src/compiler/diagnosticMessages.json]: These also?

* [line 818](https://github.com/Microsoft/TypeScript/blob/master/src/compiler/diagnosticMessages.json#L818)
* [line 1190](https://github.com/Microsoft/TypeScript/blob/master/src/compiler/diagnosticMessages.json#L1190)
* [line 1198](https://github.com/Microsoft/TypeScript/blob/master/src/compiler/diagnosticMessages.json#L1198)
* [line 1206](https://github.com/Microsoft/TypeScript/blob/master/src/compiler/diagnosticMessages.json#L1206)

If you want to change the various messages we use for these cases, I'd rather we do that in a different PR.

Comment (rbuckton) [inline src/compiler/diagnosticMessages.json]: Also "when using async functions" is too broad. It only enforces this check if you use async functions in the current module, hence the wording: "in the top level scope of a module containing async functions."

File: src/compiler/checker.ts (hunk 8 above)
Comment (yortus) [inline src/compiler/checker.ts]: Can you give a quick example where checking assignability instead of exactly matching `Promise<T>` will be a problem, either now or under future downlevelling plans?

Comment (yortus) [inline src/compiler/checker.ts]: ES3/5 restrictions need not be applied to ES6 code. Why not this:

```typescript
if (languageVersion >= ScriptTarget.ES6) {
    // use checkTypeAssignableTo to check return type annotation
}
else {
    // more restrictive return type annotation rules for ES3/5...
}
```

Comment (yortus) [inline src/compiler/checker.ts]: @mhegazy you gave qualitied support [back in December](https://github.com/Microsoft/TypeScript/issues/5911#issuecomment-162979457) for treating return type annotations one way for ES3/5 (improve the error message), and another way for ES6+ (use `checkTypeAssignableTo`). Is this still the case?

Comment (rbuckton) [inline src/compiler/checker.ts]: The current support we have in TypeScript 1.7.x for async functions allows you to specify a custom Promise implementation in the return type annotation. This was intended to support runtimes that do not yet define a global Promise implementation, and was primarily intended to support down-level emit for async functions in ES5, which is a goal for TypeScript 2.0.  This allows you to write an async function in the following fashion:

```ts
import * as bluebird from 'bluebird';

export async function fn(): bluebird.Promise {
}

console.log(fn() instanceof bluebird.Promise); // true 
```

However, this approach is not compatible with the ES7 approach to async functions which only leverages the native Promise implementation. As this native Promise is specified as part of ES6, we have elected to only allow you to supply a global `Promise<T>` return type for ES6, but reserve the previous functionality to eventually support async functions down-level.

If we were to then allow you to specify the return type of an async function as any type to which `Promise<T>` is assignable, then the above sample would still compile successfully, due to how "bluebird.d.ts" is currently implemented on DefinitelyTyped. The result is that upgrading from TypeScript 1.7.x to TypeScript 1.8 would result in completely different runtime semantics with *no indication to the developer* that anything had changed.

```ts
import * as bluebird from 'bluebird';

export async function fn(): bluebird.Promise {
}

console.log(fn() instanceof bluebird.Promise); // false?! 
```

By restricting the return type annotation to **only** the global `Promise<T>` type, we have the ability to loosen this restriction in the future. Yes, this does not fully align with how we support type annotations on generator functions; however, this is the only reliable way forward. We may, in a future release, opt to add a different mechanism for defining a compatible Promise implementation for ES5 async functions. If that does indeed become the case, we would be able to loosen the return type restriction. 

As this is a breaking change from TypeScript 1.7 behavior, it also is best to be more conservative for one to two releases to ensure any existing implementations have adjusted to this change before examining the feasibility of relaxing this restriction.

File: src/compiler/diagnosticMessages.json (hunk 7 above)
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: Let's just do that in a different PR like you mentioned.

**Comment by DanielRosenwasser** [comment]
:+1:

File: src/compiler/checker.ts (hunk 8 above)
Comment (jods4) [inline src/compiler/checker.ts]: The fact that the return type must be strictly equal to `Promise` might become a hindrance in scenarios involving interfaces and/or methods overriding.

You can work around it by adding a layer of indirection (like all CS problems :wink:) but it doesn't feel nice.

Here's one example: in frameworks such as Aurelia, many methods can return a value or a Promise for a value. Not mandating a Promise is both convenient when implementing trivial methods (like `canSave() { return true; }`) and good for perf.

Say I have a base class that exposes such a function and is meant to be overriden. Assume that the base default implementation is async.

The natural way is this:
```typescript
class BaseViewModel {
  async isValid() : boolean | Promise<boolean> { 
    /* some async implementation, maybe server call */  
  }
}

class DerivedViewModel extends BaseViewModel {
  isValid() : boolean | Promise<boolean> {
    return true;  // Non-async implementation
  }
}
```

but this wouldn't work :(

Comment (yortus) [inline src/compiler/checker.ts]: @jods4 good real world example. Current behaviour:

```typescript
var q: () => boolean | Promise<boolean>;
q = () => true;                                          // OK
q = async () => true;                                    // OK
(): boolean | Promise<boolean> => true;                  // OK
(): boolean | Promise<boolean> => Promise.resolve(true); // OK
async (): boolean | Promise<boolean> => true;            // ERROR
```

Comment (yortus) [inline src/compiler/checker.ts]: @jods4 since this PR is done and closed may I suggest taking your example across to #6686 where I think it would be quite relevent.
